File: scripts/encryptEllipticCurve.js
const crypto = require('crypto');
const keyDerivate = require('./keyDerivate.js');

const debug = false;


function encrypt(privateKeyEncryptor, publicKeyDecryptor, plaintext) {
    if (debug) console.log("Encrypting...");

    const derivedKey = keyDerivate.deriveKey(privateKeyEncryptor, publicKeyDecryptor);

    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv('aes-256-gcm', derivedKey, iv);
    const ciphertext = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
    const tag = cipher.getAuthTag();

    const ivBase64 = iv.toString('base64');
    const tagBase64 = tag.toString('base64');
    const ciphertextBase64 = ciphertext.toString('base64');

    return {
        iv: ivBase64,
        tag: tagBase64,
        ciphertext: ciphertextBase64
    };
}

function decrypt(privateKeyDecryptor, publicKeyEncryptor, encryptedData) {
    if (debug) console.log("Decrypting...");

    const derivedKey = keyDerivate.deriveKey(privateKeyDecryptor, publicKeyEncryptor);

    const iv = Buffer.from(encryptedData.iv, 'base64');
    const tag = Buffer.from(encryptedData.tag, 'base64');
    const ciphertext = Buffer.from(encryptedData.ciphertext, 'base64');

    const decipher = crypto.createDecipheriv('aes-256-gcm', derivedKey, iv);
    decipher.setAuthTag(tag);

    let plaintext;
    try {
        plaintext = Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString('utf8');
    } catch (error) {
        throw new Error("Authentication failed. Invalid ciphertext or key.");
    }

    return plaintext;
}

if (require.main === module) {
    const keyGenerate = require('./keyGenerate');
  
    const aliceKeyPair = keyGenerate.generateKeyPair();
    const bobKeyPair = keyGenerate.generateKeyPair();

    const plaintext = "Hello World!";
    const ciphertext = encrypt(aliceKeyPair.privateKey, bobKeyPair.publicKey, plaintext);
    const decryptedPlaintext = decrypt(bobKeyPair.privateKey, aliceKeyPair.publicKey, ciphertext);

    if (debug) {
        console.log();
        console.log("plaintext:           ", plaintext);
        console.log("ciphertext:          ", ciphertext);
        console.log("decryptedPlaintext: ", decryptedPlaintext);
        console.log();
    }

    if (plaintext === decryptedPlaintext) {
        console.log("All tests passed.");
    } else {
        console.log("Test failed.");
    }
}

module.exports = {
    encrypt,
    decrypt
};

File: scripts/encrypt_elliptic_curve.py
import secrets
import key_generate
import key_derivate
import base64

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.exceptions import InvalidTag
from cryptography.hazmat.primitives.asymmetric import ec

debug: bool = False


def encrypt(private_key_encryptor: ec.EllipticCurvePrivateKey, public_key_decryptor: ec.EllipticCurvePublicKey, plaintext: bytes) -> dict:
    if debug: print("Encrypting...")

    # Generate a symmetric key from public and private keys of both parties
    derived_key = key_derivate.derive_key(private_key_encryptor, public_key_decryptor)

    # Select algorythm and mode for symmetric encryption
    iv = secrets.token_bytes(16)
    cipher = Cipher(algorithms.AES(derived_key), modes.GCM(iv), backend=default_backend())
    encryptor = cipher.encryptor()

    # Encrypt the plaintext and get the associated ciphertext
    ciphertext = encryptor.update(plaintext) + encryptor.finalize()
    tag = encryptor.tag

    iv_base64 = base64.b64encode(iv).decode()
    tag_base64 = base64.b64encode(tag).decode()
    ciphertext_base64 = base64.b64encode(ciphertext).decode()

    return {
        'iv': iv_base64,
        'tag': tag_base64,
        'ciphertext': ciphertext_base64
    }

def decrypt(private_key_decryptor: ec.EllipticCurvePrivateKey, public_key_encryptor: ec.EllipticCurvePublicKey, encrypted_data: dict) -> bytes:
    if debug: print("Decrypting...")

    # Generate a symmetric key from public and private keys of both parties
    derived_key = key_derivate.derive_key(private_key_decryptor, public_key_encryptor)

    # Decode the base64 encoded attributes
    try: 
        iv = base64.b64decode(encrypted_data['iv'])
        tag = base64.b64decode(encrypted_data['tag'])
        ciphertext = base64.b64decode(encrypted_data['ciphertext'])
    except KeyError:
        raise ValueError("Invalid encrypted data format. Expected a json with keys 'iv', 'tag' and 'ciphertext, but got" + encrypted_data.keys().join(", "))

    cipher = Cipher(algorithms.AES(derived_key), modes.GCM(iv, tag), backend=default_backend())
    decryptor = cipher.decryptor()

    # Decrypt the ciphertext
    try:
        plaintext = decryptor.update(ciphertext) + decryptor.finalize()
    except InvalidTag:
        raise ValueError("Authentication failed. Invalid ciphertext or key.")

    return plaintext


if __name__ == "__main__":
    alice_private_key, alice_public_key = key_generate.generate_key_pair()
    bob_private_key, bob_public_key = key_generate.generate_key_pair()

    plaintext = "Hello World!"
    ciphertext = encrypt(alice_private_key, bob_public_key, plaintext.encode())
    decrypted_plaintext = decrypt(bob_private_key, alice_public_key, ciphertext)
    decrypted_plaintext = decrypted_plaintext.decode()

    if debug: 
        print()
        print("plaintext:           ", plaintext)
        print("ciphertext:          ", ciphertext)
        print("decrypted_plaintext: ", decrypted_plaintext)
        print()

    assert plaintext == decrypted_plaintext

    print("All tests passed.")

